### PR TITLE
always set lastTransitionTime when crd condition changes

### LINKED ISSUE #69654: CRD Established condition has null for lastTransitionTime
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:  the `lastTransitionTime` for a CRDs Established condition is always null

**What you expected to happen**: it should have a timestamp

**How to reproduce it (as minimally and precisely as possible)**: create a CRD and then examine its yaml/json. For example:

```json
"status": {
                "acceptedNames": {
                    "kind": "Foo",
                    "listKind": "FooList",
                    "plural": "foos",
                    "singular": "foo"
                },
                "conditions": [
                    {
                        "lastTransitionTime": "2018-10-10T11:28:11Z",
                        "message": "no conflicts found",
                        "reason": "NoConflicts",
                        "status": "True",
                        "type": "NamesAccepted"
                    },
                    {
                        "lastTransitionTime": null,
                        "message": "the initial names have been accepted",
                        "reason": "InitialNamesAccepted",
                        "status": "True",
                        "type": "Established"
                    }
                ],
                "storedVersions": [
                    "v1alpha1"
                ]
            }
```

**Anything else we need to know?**:

**Environment**:
community master

### PR DESCRIPTION
**Pull request by CaoShuFeng**
Fixes: https://github.com/kubernetes/kubernetes/issues/69654

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
When the status of a CRD Condition changes, it's corresponding `LastTransitionTime` is now updated.
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ func SetCRDCondition(crd *CustomResourceDefinition, newCondition CustomResourceD
 
 	if existingCondition.Status != newCondition.Status {
 		existingCondition.Status = newCondition.Status
-		existingCondition.LastTransitionTime = newCondition.LastTransitionTime
+		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: Wow, thanks for catching this, @CaoShuFeng!

I think we can make it cleaner by writing `newCondition.LastTransitionTime = metav1.NewTime(time.Now())` on the top of the func instead. Wdyt?

```diff
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go
@@ -26,9 +26,9 @@ import (
 // SetCRDCondition sets the status condition.  It either overwrites the existing one or
 // creates a new one
 func SetCRDCondition(crd *CustomResourceDefinition, newCondition CustomResourceDefinitionCondition) {
+       newCondition.LastTransitionTime = metav1.NewTime(time.Now())
        existingCondition := FindCRDCondition(crd, newCondition.Type)
        if existingCondition == nil {
-               newCondition.LastTransitionTime = metav1.NewTime(time.Now())
                crd.Status.Conditions = append(crd.Status.Conditions, newCondition)
                return
        }
```

Comment (CaoShuFeng) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: Done

**Comment by CaoShuFeng** [comment]
> I think it'd be nice to have a release note for this. Something like:

Done.

**Comment by nikhita** [comment]
Thanks @CaoShuFeng! 

/lgtm

/assign @sttts 
for approval

**Comment by nikhita** [comment]
/lgtm cancel

It looks good, but let's also add a test for it?

https://github.com/kubernetes/kubernetes/blob/effd009dabcd0738f5e2b0112b2d3b943c842c69/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers_test.go#L93

**Comment by CaoShuFeng** [comment]
>It looks good, but let's also add a test for it?

Done.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers_test.go [hunk 2]
```diff
@@ -202,6 +202,67 @@ func TestSetCRDCondition(t *testing.T) {
...
+					Reason:             "Accepted",
+					Message:            "the initial names have been accepted",
+					LastTransitionTime: metav1.Date(2018, 1, 1, 0, 0, 0, 0, time.UTC),
+				},
+				{
+					Type:               Terminating,
+					Status:             ConditionFalse,
+					Reason:             "NeverEstablished",
+					Message:            "resource was never established",
+					LastTransitionTime: metav1.Date(2018, 2, 1, 0, 0, 0, 0, time.UTC),
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers_test.go]: I'm just curious how this works...shouldn't `metav1.NewTime(time.Now())` give the current local time? How is it `metav1.Date(2018, 2, 1, 0, 0, 0, 0, time.UTC)` everytime?

**Review by nikhita — COMMENTED** [review]
Lgtm, just one question for my understanding.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers_test.go (hunk 2 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers_test.go]: Sorry, somehow missed this reply. :(

**Comment by nikhita** [comment]
/lgtm
/priority important-soon

@sttts can you approve?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go [hunk 3]
```diff
@@ -26,9 +26,9 @@ import (
 // SetCRDCondition sets the status condition.  It either overwrites the existing one or
 // creates a new one
 func SetCRDCondition(crd *CustomResourceDefinition, newCondition CustomResourceDefinitionCondition) {
+	newCondition.LastTransitionTime = metav1.NewTime(time.Now())
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: shouldn't we have some deepequal comparison here?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: I don't follow...why do we need to compare here?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: the timestamp should not change if the condition being set already exists and has the same status

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go [hunk 4]
```diff
@@ -26,9 +26,9 @@ import (
...
 	existingCondition := FindCRDCondition(crd, newCondition.Type)
 	if existingCondition == nil {
-		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
 		crd.Status.Conditions = append(crd.Status.Conditions, newCondition)
 		return
 	}
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: if we have an existing condition with a zero LastTransitionTime, we also need to update:
```go
if existingCondition.Status != newCondition.Status || existingCondition.LastTransitionTime.IsZero() {
```

**Comment by liggitt** [comment]
one change needed (and tests updated accordingly), then lgtm

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: done

**Comment by nikhita** [comment]
/lgtm

/assign @liggitt 
for approval

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go [hunk 5]
```diff
@@ -27,20 +27,21 @@ var swaggerMetadataDescriptions = metav1.ObjectMeta{}.SwaggerDoc()
...
-	if existingCondition.Status != newCondition.Status {
-		existingCondition.Status = newCondition.Status
-		existingCondition.LastTransitionTime = newCondition.LastTransitionTime
-	}
-
+	existingCondition.Status = newCondition.Status
 	existingCondition.Reason = newCondition.Reason
 	existingCondition.Message = newCondition.Message
+
+	if existingCondition.Status != newCondition.Status || existingCondition.LastTransitionTime.IsZero() {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: switching the order breaks the condition.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go]: fixed

**Pull request merged**